src/shared/serialize.c

```diff
@@ -116,6 +116,26 @@ int serialize_fd(FILE *f, FDSet *fds, const char *key, int fd) {
         return serialize_item_format(f, key, "%i", copy);
 }
 
+int serialize_fd_many(FILE *f, FDSet *fds, const char *key, const int fd_array[], size_t n_fd_array) {
+        _cleanup_free_ char *t = NULL;
+
+        assert(f);
+        assert(fd_array || n_fd_array == 0);
+
+        for (size_t i = 0; i < n_fd_array; i++) {
+                int copy;
+
+                copy = fdset_put_dup(fds, fd_array[i]);
```

Comment:
the fdset_put_dup should be done only if fd_array[i] is valid (eg: the pipe_fd array can contain -EBADF at some positions), and copy initialized to -EBADF

Reply:
pipe_fd afaics is either fully valid or fully invalid if the system is working. but yeah, we need to check for this case. will add.

Reply:
as mentioned in the other PR I don't have any references, but some vague memories that in some cases it wasn't fully initialized